File: 57.inset_interval/insert.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        #init
        if not intervals:
            return [newInterval]
        if not newInterval:
            return intervals
        res = []
        pos = 0
        #traverse the list, merge if overlapping
        for inter in intervals:
            if inter.end < newInterval.start:
                res.append(inter)
                pos += 1
            elif inter.start > newInterval.end:
                res.append(inter)
            else:
                newInterval.start = min(inter.start, newInterval.start)
                newInterval.end = max(inter.end, newInterval.end)
        res.insert(pos, newInterval)
        return res


#time: O(n)
#space: O(1)

File: 57.inset_interval/insert.js
/**
 * Given a set of non-overlapping intervals, insert a new interval into the intervals (merge if necessary).
 * You may assume that the intervals were initially sorted according to their start times.
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */

// loop through the array and find the lower bound and higher bound
// then merge the interval
// the difficult part is handling the insert, should only increase the insert index
// when the newInterval is behind the interval traversed
var insert = function(intervals, newInterval) {
  const newArray = [];
  let insertIndex = 0;
  for (let i = 0; i < intervals.length; i++) {
    // no overlap
    if (newInterval[0] > intervals[i][1]) {
      newArray.push(intervals[i]);
      insertIndex++;
    } else if (newInterval[1] < intervals[i][0]) {
      newArray.push(intervals[i]);
    } else {
      newInterval = [
        Math.min(newInterval[0], intervals[i][0]),
        Math.max(newInterval[1], intervals[i][1]),
      ];
    }
  }
  newArray.splice(insertIndex, 0, newInterval);
  return newArray;
};
